### ISSUE TITLE
[Cleanup] Unused Variable Declarations - Code quality cleanup

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#3011]
## Summary
Multiple unused variable declarations exist across the codebase marked with FIXME(unused) comments.

## Files Affected
- `shared/training/loops/training_loop.mojo:108,244` - Unused accuracy_metric and actual_batch_size
- `shared/testing/layer_testers.mojo:1128-1129` - Unused epsilon and tolerance
- `tests/shared/core/test_normalization.mojo:329-330` - Unused grad_gamma and grad_beta
- `tests/shared/core/test_dropout.mojo:173,277,353` - Unused output and mask variables
- `tests/shared/core/test_nvfp4.mojo:166,177,185` - Unused conversion results

## Current State
Variables are declared but not used, with FIXME(unused) comments indicating awareness.

## Options
1. Use the variables for their intended purpose
2. Remove the unused declarations entirely
3. Convert to test assertions if they should be verified

## Acceptance Criteria
- [ ] Review each unused variable for intent
- [ ] Either use or remove each variable
- [ ] Remove all FIXME(unused) comments

## Previous Issue
Replaces closed #2710

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Now I have all the context I need. Let me analyze each unused variable and determine the appropriate action.

# Implementation Plan: Issue #3011 - Unused Variable Declarations Cleanup

## Summary

This issue addresses code quality by resolving 13 unused variable declarations across 5 files. After analyzing each case, the variables fall into three categories:

1. **Test result captures that can be removed** - Variables that capture function results but are never validated
2. **Placeholder variables for future implementation** - Variables reserved for features not yet implemented  
3. **Intentionally unused backward pass results** - Gradient outputs not used in specific tests

The recommended approach is to **remove all unused variables** since:
- Test captures without assertions provide no testing value
- Placeholder code should be added when actually needed (YAGNI principle)
- The FIXME comments already document these are intentional omissions

---

## Step-by-Step Implementation Tasks

### Task 1: Clean up `shared/training/loops/training_loop.mojo`

**File**: `/home/mvillmow/ProjectOdysseyManual/shared/training/loops/training_loop.mojo`

**Location 1: Line 108** - `accuracy_metric`
```mojo
# FIXME(unused) var accuracy_metric = AccuracyMetric()
```
- **Analysis**: This creates an `AccuracyMetric` instance but never uses it. The function currently passes `0.0` as a placeholder for accuracy at line 150.
- **Decision**: **REMOVE** - The accuracy tracking feature is not implemented. The import can remain for future use, but the instantiation should be removed until the feature is built.
- **Action**: Delete line 108 entirely.

**Location 2: Line 244** - `actual_batch_size`
```mojo
# FIXME(unused) var actual_batch_size = end_idx - start_idx
```
- **Analysis**: Calculates actual batch size (which differs from `batch_size` at end of dataset), but this value is never used for anything.
- **Decision**: **REMOVE** - If padding/normalization for the last batch is needed, it should be added with actual implementation. Currently the batch slicing works correctly without this variable.
- **Action**: Delete line 244 entirely.

---

### Task 2: Clean up `shared/testing/layer_testers.mojo`

**File**: `/home/mvillmow/ProjectOdysseyManual/shared/testing/layer_testers.mojo`

**Location: Lines 1128-1129** - `epsilon` and `tolerance`
```mojo
# FIXME(#2710, unused) var epsilon = 1e-5 if dtype == DType.float32 else 1e-4
# FIXME(#2710, unused) var tolerance = 1e-2 if dtype == DType.float32 else 1e-1
```
- **Analysis**: These are declared in `test_batchnorm_layer_backward()` for gradient checking, but the actual gradient checking code is not yet implemented (see comment at line 1131: "Note: Actual BatchNorm backward gradient checking would be implemented when BatchNorm forward pass is available").
- **Decision**: **REMOVE** - This is placeholder code for future gradient checking. The note already documents what needs to happen when implementing. YAGNI - add these when the feature is built.
- **Action**: Delete lines 1128-1129 entirely.

---

### Task 3: Clean up `tests/shared/core/test_normalization.mojo`

**File**: `/home/mvillmow/ProjectOdysseyManual/tests/shared/core/test_normalization.mojo`

**Location: Lines 329-330** - `grad_gamma` and `grad_beta`
```mojo
var grad_input = result7[0]
# FIXME(unused) var grad_gamma = result7[1]
# FIXME(unused) var grad_beta = result7[2]
```
- **Analysis**: In `test_batch_norm2d_backward_gradient_input()`, the function tests gradient w.r.t. input only. The `grad_gamma` and `grad_beta` values are returned but intentionally not tested in this specific test.
- **Decision**: **REMOVE** - The test name clearly specifies it's testing "gradient_input". Separate tests should be added for `grad_gamma` and `grad_beta` if needed. Capturing unused values just to comment them out adds noise.
- **Action**: Delete lines 329-330 entirely.

---

### Task 4: Clean up `tests/shared/core/test_dropout.mojo`

**File**: `/home/mvillmow/ProjectOdysseyManual/tests/shared/core/test_dropout.mojo`

**Location 1: Line 173** - `output` in `test_dropout_backward_shapes()`
```mojo
# FIXME(unused) var output = result7[0]
var mask = result7[1]
```
- **Analysis**: The test is checking backward shapes. It needs the mask for backward pass but doesn't need to verify output.
- **Decision**: **REMOVE** - The test only cares about the mask for the backward pass test.
- **Action**: Delete line 173 entirely.

**Location 2: Line 277** - `mask` in `test_dropout2d_shapes()`
```mojo
var output = result10[0]
# FIXME(unused) var mask = result10[1]
```
- **Analysis**: This test validates output shapes only, mask is not used.
- **Decision**: **REMOVE** - The test is about output shapes only.
- **Action**: Delete line 277 entirely.

**Location 3: Line 353** - `output` in `test_dropout2d_backward_shapes()`
```mojo
# FIXME(unused) var output = result13[0]
var mask = result13[1]
```
- **Analysis**: Similar to Location 1 - backward shape test that needs mask only.
- **Decision**: **REMOVE** - The test only needs the mask.
- **Action**: Delete line 353 entirely.

---

### Task 5: Clean up `tests/shared/core/test_nvfp4.mojo`

**File**: `/home/mvillmow/ProjectOdysseyManual/tests/shared/core/test_nvfp4.mojo`

**Location 1: Line 166** - `result` in `test_nvfp4_special_values_nan()`
```mojo
# FIXME(unused) var result = nvfp4_nan.to_float32()
```
- **Analysis**: The test verifies that NaN encodes to the correct bit pattern (0b0111), but doesn't verify round-trip.
- **Decision**: **REMOVE** - The test's purpose is to verify encoding, not round-trip behavior for NaN.
- **Action**: Delete line 166 entirely.

**Location 2: Line 177** - `result_pos` in `test_nvfp4_special_values_inf()`
```mojo
# FIXME(unused) var result_pos = nvfp4_pos_inf.to_float32()
```
- **Analysis**: Same pattern - tests encoding of +Inf to bit pattern.
- **Decision**: **REMOVE** - Testing encoding only.
- **Action**: Delete line 177 entirely.

**Location 3: Line 185** - `result_neg` in same function
```mojo
# FIXME(unused) var result_neg = nvfp4_neg_inf.to_float32()
```
- **Analysis**: Same pattern - tests encoding of -Inf to bit pattern.
- **Decision**: **REMOVE** - Testing encoding only.
- **Action**: Delete line 185 entirely.

---

## Files to Modify

| File | Lines to Remove | Variables |
|------|-----------------|-----------|
| `shared/training/loops/training_loop.mojo` | 108, 244 | `accuracy_metric`, `actual_batch_size` |
| `shared/testing/layer_testers.mojo` | 1128, 1129 | `epsilon`, `tolerance` |
| `tests/shared/core/test_normalization.mojo` | 329, 330 | `grad_gamma`, `grad_beta` |
| `tests/shared/core/test_dropout.mojo` | 173, 277, 353 | `output`, `mask`, `output` |
| `tests/shared/core/test_nvfp4.mojo` | 166, 177, 185 | `result`, `result_pos`, `result_neg` |

**Total**: 13 unused variable declarations to remove across 5 files.

---

## Testing Approach

### Pre-Implementation Verification
1. Run existing test suite to establish baseline:
   ```bash
   just test-mojo tests/shared/core/test_normalization.mojo
   just test-mojo tests/shared/core/test_dropout.mojo
   just test-mojo tests/shared/core/test_nvfp4.mojo
   ```

2. Build the shared library to verify no compile errors:
   ```bash
   just build
   ```

### Post-Implementation Verification
1. **Compile verification**: All files should compile without warnings:
   ```bash
   just build
   ```

2. **Unit test verification**: Run affected test files:
   ```bash
   just test-mojo tests/shared/core/test_normalization.mojo
   just test-mojo tests/shared/core/test_dropout.mojo
   just test-mojo tests/shared/core/test_nvfp4.mojo
   ```

3. **Full test suite**: Run comprehensive tests to catch any regressions:
   ```bash
   just test
   ```

4. **Pre-commit hooks**: Ensure code passes linting:
   ```bash
   just pre-commit-all
   ```

### Verification Checklist
- [ ] All 5 files compile successfully
- [ ] No new compiler warnings introduced
- [ ] All existing tests pass
- [ ] No FIXME(unused) comments remain in modified files
- [ ] Pre-commit hooks pass

---

## Success Criteria

- [x] All 13 `FIXME(unused)` variable declarations removed
- [ ] All 5 files compile without errors
- [ ] All affected test files pass
- [ ] Full test suite passes (`just test`)
- [ ] Pre-commit hooks pass (`just pre-commit-all`)
- [ ] No functional regressions (tests still validate the same behavior)
- [ ] Code follows minimal changes principle (only remove, don't refactor)

---

## Implementation Notes

### Order of Operations
1. Create feature branch: `git checkout -b 3011-cleanup-unused-variables`
2. Make edits to each file (simple line deletions)
3. Run `just build` after each file to verify compilation
4. Run affected test files after all edits
5. Run `just pre-commit-all` for final verification
6. Create PR with `gh pr create --body "Closes #3011"`

### Risk Assessment
- **Risk Level**: Low
- **Rationale**: All changes are simple line deletions of commented-out or dead code
- **Rollback**: Easy - can revert any individual file change

### Dependencies
- No external dependencies
- No changes to public APIs
- No changes to test assertions or behavior

---

## Resource Usage

- **Files Read**: 5
- **Tool Calls Used**: 5 (all Read operations)
- **Remaining Budget**: 45 tool calls, 49 steps
